### PR TITLE
[All] Page embedding

### PR DESCRIPTION
**Pull request by hartez**
### Description of Change ###

This change allows a developer to navigate to a page expressed as a Xamarin.Forms XAML ContentPage in a Xamarin iOS, Android or UWP application.

In addition to enabling this feature, this change also adds Android, UWP, and iOS native test bed projects for this feature; in the future these can be integrated into our full UI test suite.

Most of the API changes are in support of Android applications which need to support Forms pages in multiple Activities. The interesting additions are:

- `public static UIViewController CreateViewController(this ContentPage page)` for wrapping a Forms ContentPage in a UIViewController which can be used with a UINavigationController on iOS
- `public static FrameworkElement CreateFrameworkElement(this ContentPage contentPage)` for wrapping a Forms ContentPage in a FrameworkElement which can be used in a UWP application
- `public static bool Navigate(this Windows.UI.Xaml.Controls.Frame frame, ContentPage page)` which allows a Forms ContentPage to be used with UWP Frame navigation
- `public static Fragment CreateFragment(this ContentPage view, Context context)` and `public static Android.Support.V4.App.Fragment CreateSupportFragment(this ContentPage view, Context context)`, which allow a ContentPage to be wrapped in a Fragment or support Fragment for navigation in an Android application.

#### Added:

Xamarin.Forms.Internals.Registrar
- `public TOut GetHandler<TOut>(Type type, params object[] args)`

Xamarin.Forms.Platform.UWP
- `public static FrameworkElement CreateFrameworkElement(this ContentPage contentPage)`
- `public static bool Navigate(this Windows.UI.Xaml.Controls.Frame frame, ContentPage page)`

Xamarin.Forms.Platform.iOS
- `public static UIViewController CreateViewController(this ContentPage page)`

Xamarin.Forms
- `public static void SetTitleBarVisibility(Activity activity, AndroidTitleBarVisibility visibility)`

Xamarin.Forms.Maps.Android
- `public MapRenderer(Context context)`

Xamarin.Forms.Platform.Android
- `public static Fragment CreateFragment(this ContentPage view, Context context)`
- `public static Android.Support.V4.App.Fragment CreateSupportFragment(this ContentPage view, Context context)`
- `ActivityIndicatorRenderer(Context)`
- `BoxRenderer(Context)`
- `ButtonRenderer(Context)`
- `CarouselPageRenderer(Context)`
- `public static ColorExtensions.ToAndroid(this Color, int, Context)`
- `DatePickerRenderer(Context)`
- `EditorRenderer(Context)`
- `EntryRenderer(Context)`
- `FrameRenderer(Context)`
- `ImageRenderer(Context)`
- `public InnerScaleListener(Func<float, Point, bool> pinchDelegate, Func<Point, bool> pinchStarted, Action pinchEnded, Func<double, double> fromPixels)`
- `LabelRenderer(Context)`
- `ListViewRenderer(Context)`
- `MasterDetailRenderer(Context)`
- `NativeViewWrapperRenderer(Context)`
- `NavigationMenuRenderer(Context)`
- `NavigationRenderer(Context)`
- `OpenGLViewRenderer(Context)`
- `PageRenderer(Context)`
- `PickerRenderer(Context)`
- `public static Platform.CreateRenderer(VisualElement, Context)`
- `ProgressBarRenderer(Context)`
- `public static ResourceManager.GetDrawable(this Context, string)`
- `ScrollViewRenderer(Context)`
- `SearchBarRenderer(Context)`
- `SliderRenderer(Context)`
- `StepperRenderer(Context)`
- `SwitchRenderer(Context)`
- `TabbedRenderer(Context)`
- `TableViewRenderer(Context)`
- `TimePickerRenderer(Context)`
- `ViewRenderer(Context)`
- `VisualElementRenderer(Context)`
- `WebViewRenderer(Context)`
- `DefaultRenderer(Context)`

Xamarin.Forms.Platform.Android.AppCompat
- `ButtonRenderer(Context)`
- `CarouselPageRenderer(Context)`
- `FrameRenderer(Context)`
- `MasterDetailPageRenderer(Context)`
- `NavigationPageRenderer(Context)`
- `PickerRenderer(Context)`
- `SwitchRenderer(Context)`
- `TabbedPageRenderer(Context)`
- `ViewRenderer(Context)`

Xamarin.Forms.Platform.Android.FastRenderers
- `ButtonRenderer(Context)`
- `FrameRenderer(Context)`
- `ImageRenderer(Context)`
- `LabelRenderer(Context)`

#### Obsoleted: 

Xamarin
- `public static FormsMaps.Context`

Xamarin.Forms
- `public static Forms.Context`
- `public static Forms.SetTitleBarVisibility(AndroidTitleBarVisibility visibility)`

Xamarin.Forms.Maps.Android
- `MapRenderer.MapRenderer()`

Xamarin.Forms.Platform.Android
- `ActivityIndicatorRenderer.ActivityIndicatorRenderer()`
- `BoxRenderer.BoxRenderer()`
- `ButtonRenderer.ButtonRenderer()`
- `CarouselPageRenderer.CarouselPageRenderer()`
- `ColorExtensions.ToAndroid(this Color self, int defaultColorResourceId):Color`
- `DatePickerRenderer.DatePickerRenderer()`
- `EditorRenderer.EditorRenderer()`
- `EntryRenderer.EntryRenderer()`
- `FrameRenderer.FrameRenderer()`
- `ImageRenderer.ImageRenderer()`
- `InnerScaleListener.InnerScaleListener(Func<float,Point,bool> pinchDelegate, Func<Point,bool> pinchStarted, Action pinchEnded)`
- `LabelRenderer.LabelRenderer()`
- `ListViewRenderer.ListViewRenderer()`
- `MasterDetailRenderer.MasterDetailRenderer()`
- `NativeViewWrapperRenderer.NativeViewWrapperRenderer()`
- `NavigationMenuRenderer.NavigationMenuRenderer()`
- `NavigationRenderer.NavigationRenderer()`
- `OpenGLViewRenderer.OpenGLViewRenderer()`
- `PageRenderer.PageRenderer()`
- `PickerRenderer.PickerRenderer()`
- `Platform.CreateRenderer(VisualElement element):IVisualElementRenderer`
- `ProgressBarRenderer.ProgressBarRenderer()`
- `ResourceManager.GetDrawable(this Resources resource, string name):Drawable`
- `ScrollViewRenderer.ScrollViewRenderer()`
- `SearchBarRenderer.SearchBarRenderer()`
- `SliderRenderer.SliderRenderer()`
- `StepperRenderer.StepperRenderer()`
- `SwitchRenderer.SwitchRenderer()`
- `TabbedRenderer.TabbedRenderer()`
- `TableViewRenderer.TableViewRenderer()`
- `TimePickerRenderer.TimePickerRenderer()`
- `ViewRenderer.ViewRenderer()`
- `ViewRenderer<TView,TNativeView>.ViewRenderer()`
- `VisualElementRenderer<TElement>.VisualElementRenderer()`
- `WebViewRenderer.WebViewRenderer()`
- `Xamarin.Forms.Platform.Android.Platform.DefaultRenderer.DefaultRenderer()`

Xamarin.Forms.Platform.Android.AppCompat
- `ButtonRenderer.ButtonRenderer()`
- `CarouselPageRenderer.CarouselPageRenderer()`
- `FrameRenderer.FrameRenderer()`
- `MasterDetailPageRenderer.MasterDetailPageRenderer()`
- `NavigationPageRenderer.NavigationPageRenderer()`
- `PickerRenderer.PickerRenderer()`
- `SwitchRenderer.SwitchRenderer()`
- `TabbedPageRenderer.TabbedPageRenderer()`
- `ViewRenderer<TView,TControl>.ViewRenderer()`
	  
Xamarin.Forms.Platform.Android.FastRenderers
- `ButtonRenderer.ButtonRenderer()`
- `FrameRenderer.FrameRenderer()`
- `ImageRenderer.ImageRenderer()`
- `LabelRenderer.LabelRenderer()`
 
### Behavioral Changes ###

Users who have subclassed Android renderers will need to update their code to use the new constructors or they will see Obsolete warnings.

### PR Checklist ###

- [ ] Has tests (if omitted, state reason in description)
- [x] Rebased on top of master at time of PR
- [x] Changes adhere to coding standard
- [x] Consolidate commits as makes sense

### CODE REVIEW

File: EmbeddingTestBeds/Embedding.Droid/Assets/AboutAssets.txt [hunk 1]
```diff
@@ -0,0 +1,19 @@
+Any raw assets you want to be deployed with your application can be placed in
+this directory (and child directories) and given a Build Action of "AndroidAsset".
+
+These files will be deployed with you package and will be accessible using Android's
```
Comment (samhouts) [inline EmbeddingTestBeds/Embedding.Droid/Assets/AboutAssets.txt]: Someone needs to fix that typo. `with you package`.

File: Xamarin.Forms.Platform.Android/ButtonBackgroundTracker.cs [hunk 2]
```diff
@@ -52,7 +52,7 @@ public void UpdateDrawable()
 			else
 			{
 				if (_backgroundDrawable == null)
-					_backgroundDrawable = new ButtonDrawable();
+					_backgroundDrawable = new ButtonDrawable(_nativeButton.Context.ToPixels);
```
Comment (samhouts) [inline Xamarin.Forms.Platform.Android/ButtonBackgroundTracker.cs]: Gonna have conflicts here when #1178 is merged.

File: Xamarin.Forms.Platform.Android/Renderers/ScrollViewRenderer.cs [hunk 3]
```diff
@@ -229,10 +235,10 @@ internal void UpdateScrollPosition(double x, double y)
 				if (_view.Orientation == ScrollOrientation.Both)
 				{
 					if (x == 0)
-						x = Forms.Context.FromPixels(_hScrollView.ScrollX);
```
Comment (jassmith) [inline Xamarin.Forms.Platform.Android/Renderers/ScrollViewRenderer.cs]: Don't double fetch Context in the same method. Its a JNI call and expensive.

File: Xamarin.Forms.Platform.Android/Renderers/ScrollViewRenderer.cs [hunk 4]
```diff
@@ -219,7 +225,7 @@ protected override void OnScrollChanged(int l, int t, int oldl, int oldt)
 		{
 			base.OnScrollChanged(l, t, oldl, oldt);
 
-			UpdateScrollPosition(Forms.Context.FromPixels(l), Forms.Context.FromPixels(t));
```
Comment (jassmith) [inline Xamarin.Forms.Platform.Android/Renderers/ScrollViewRenderer.cs]: Double Context fetch

File: EmbeddingTestBeds/Embedding.Droid/Assets/AboutAssets.txt (hunk 1 above)
Comment (hartez) [reply EmbeddingTestBeds/Embedding.Droid/Assets/AboutAssets.txt]: Should be fixed now.

File: Xamarin.Forms.Platform.Android/Renderers/ScrollViewRenderer.cs (hunk 4 above)
Comment (hartez) [reply Xamarin.Forms.Platform.Android/Renderers/ScrollViewRenderer.cs]: Fixed.

**Pull request merged**